docs/concurrency_design.md

```diff
@@ -129,3 +129,90 @@ sequenceDiagram
+- Validating
+- Performing state transition.
+- Saving the state to the DB/cache
+- At the end of all, updating the global store (this is a very fast operation, contrary to state transition).
+
+With this strategy in mind, the block won’t need to be passed around during its lifetime, and state transitions may happen in parallel without the need to spawn tasks on demand.
+
+There are some tasks that, of course, will require interaction with the outside world:
+
+- When a parent is missing, we need to download it. The block processor may pause until it receives a notification that the parent was downloaded and transitioned. After that, it may continue with its own state transition.
+- When validating it will need to interact with the global fork choice store. This just requires validating a few parameters like the epoch/slot, which is a fast process, so it’s a relatively simple call without passing the full block and without changing the store state.
+- The final step is updating the store, which still doesn’t need to pass around big state and is fast. If additional validations are performed or repeated, it doesn’t represent a big performance hit.
+
+### Happy path diagram
+
+```mermaid
+sequenceDiagram
+
+participant port as LibP2P Port <br> (Genserver)
+participant bp as Block Processor (Genserver)
+participant store as Fork Choice Store <br> (Genserver)
+participant db as Persistent Store <br> (DB/Cache)
+
+port ->> bp: gossip(block)
+bp ->> bp: decompress <br> decode <br> basic validation
+bp ->> db: store_block_if_not_present(block)
+db -->> bp: true
+bp ->> db: parent_present?(block_root)
+db -->> bp: true
+bp ->> store: block_valid?(slot, parent_root)
+store -->> bp: true
+bp ->> db: state_by_block_root(parent_root)
+db -->> bp: state
+bp ->> bp: state_transition(state, block)
+bp ->> db: put_state(new_state)
+bp ->> store: new_state(block_root,<br> block_slot,<br>state_justified_checkpoint,<br>state_finalized_checkpoint)
+```
+
+We can see here that:
+
+1. We don’t need to send full blocks or states anywhere except to/from the db/cache.
+2. Most operations are independent and the interactions (e.g. the validation or the new_state) are cheap.
+
+### Missing Parent diagram
+
+```mermaid
+sequenceDiagram
+
+participant bp as Block Processor (Genserver)
+participant db as Persistent Store <br> (DB/Cache)
+participant dt as Download Task
+participant bpp as Parent Block Processor (Genserver)
+
+bp ->>+ db: parent_present?(block_root)
+activate bp
+db -->>- bp: false
+bp ->>+ dt: spawn(parent_root)
+deactivate bp
+dt ->> dt: download(parent_root)
+dt ->>- bpp: spawn(parent)
+activate bpp
+bpp ->> db: store_block_if_not_present(block)
+bpp ->> bpp: validation and state transition
+bpp ->> db: put_state(new_parent_state)
+deactivate bpp
+db ->>+ bp: parent_ready()
+bp ->>- bp: validation and state transition
+```
+
+This is a simplified sequence diagram highlighting the differences when the parent block is not present. To summarize:
```

Comment:
you mean simplified because it doesn't show the recursive nature of the problem, right?

Reply:
Yes